File: web/src/algorithms/snippets/radix.java
/** LSD radix sort — base 10 for clarity. */
public final class RadixSort {
    public static void sort(int[] data) {
        int n = data.length;
        if (n < 2) return;
        int[] a = data.clone();
        int[] b = new int[n];
        int max = 0;
        for (int v : a) if (v > max) max = v;
        int passes = Math.max(1, String.valueOf(max).length());
        for (int pass = 0; pass < passes; pass++) {
            int[] count = new int[10];
            for (int v : a) count[digit(v, pass)]++;
            int[] off = new int[10];
            int sum = 0;
            for (int d = 0; d < 10; d++) {
                off[d] = sum;
                sum += count[d];
            }
            int[] run = off.clone();
            for (int v : a) {
                int d = digit(v, pass);
                b[run[d]++] = v;
            }
            int[] t = a;
            a = b;
            b = t;
        }
        System.arraycopy(a, 0, data, 0, n);
    }

    private static int digit(int v, int pass) {
        return (v / (int) Math.pow(10, pass)) % 10;
    }
}

File: web/src/algorithms/snippets/radix.py
"""LSD radix sort, base 10."""

def radix_sort(data: list[int]) -> list[int]:
    n = len(data)
    if n < 2:
        return data[:]
    a = data[:]
    b = [0] * n
    max_v = max(a)
    passes = max(1, len(str(max_v)))
    for p in range(passes):
        count = [0] * 10
        for v in a:
            count[_digit(v, p)] += 1
        off = [0] * 10
        s = 0
        for d in range(10):
            off[d] = s
            s += count[d]
        run = off[:]
        for v in a:
            d = _digit(v, p)
            b[run[d]] = v
            run[d] += 1
        a, b = b, [0] * n
    return a


def _digit(v: int, p: int) -> int:
    return (v // (10**p)) % 10
